File: easy/876_middle_of_the_linked_list/python/solution.py
class ListNode:
    def __init__(self, value = 0, next = None): # define constructor
        self.value = value
        self.next = next

class CustomLinkedList:
    def __init__(self):
        self.head = None
        self.size = 0
    
    def insert(self, value):
        newNode = ListNode(value)
        if self.head == None:
            self.head = newNode
        else:
            tempNode = self.head
            while tempNode.next is not None:
                tempNode = tempNode.next
            tempNode.next = newNode
        self.size += 1
    
    def printList(self):
        currentNode = self.head
        print ("Linked list: ")
        while currentNode is not None:
            print(currentNode.value)
            currentNode = currentNode.next
    
    def findMiddleNode(self):
        if self.head == None:
            return None
        
        slowPtr = self.head
        fastPtr = self.head

        while fastPtr is not None and fastPtr.next is not None:
            slowPtr = slowPtr.next
            fastPtr = fastPtr.next.next
        
        return slowPtr
        

if __name__ == '__main__':
    myList = CustomLinkedList()
    myList.insert(5)
    myList.insert(15)
    myList.insert(3)
    myList.printList()
    print("Middle node is " + str(myList.findMiddleNode().value))

    myList.insert(45)
    myList.printList()
    print("Middle node is " + str(myList.findMiddleNode().value))

File: easy/876_middle_of_the_linked_list/java/solution.java


class Solution {

    public static class ListNode {
        int value;
        ListNode next;

        ListNode() {}

        ListNode(int value) {
            this.value = value;
            next = null;
        }

        ListNode(int value, ListNode next) {
            this.value = value; 
            this.next = next;
        }
    }

    public static class CustomLinkedList {
        private ListNode head;
        private int size;

        public CustomLinkedList() 
        {
            head = null;
            size = 0;
        }

        public void insert(int value)
        {
            ListNode newNode = new ListNode(value);

            if (head == null)
            {
                head = newNode;
            }
            else
            {
                ListNode tempNode = head;
                while (tempNode.next != null) // traverse the list until the last element
                {
                    tempNode = tempNode.next;
                }
                tempNode.next = newNode;
            }
            size++;
        }

        public int getSize()
        {
            return this.size;
        }

        public void printList() 
        { 
            ListNode currentNode = head; 
    
            System.out.print("LinkedList: "); 
    
            // Traverse through the LinkedList 
            while (currentNode != null) { 
                // Print the data at current node 
                System.out.print(currentNode.value + " "); 
    
                // Go to next node 
                currentNode = currentNode.next; 
            } 
        } 

        public ListNode findMiddleNode()
        {
            if (head == null)
                return null;
            
            //ListNode middleNode = new ListNode();

            ListNode slowPtr = head;
            ListNode fastPtr = head;

            //while (fastPtr != null)
            while (fastPtr != null && fastPtr.next != null)
            {
                slowPtr = slowPtr.next;
                fastPtr = fastPtr.next.next;
            }

            //middleNode = slowPtr;
            //middleNode.value = slowPtr.value;

            return slowPtr;
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        CustomLinkedList myList = new CustomLinkedList();
        myList.insert(5);
        myList.insert(15);
        myList.insert(3);
        myList.printList();
        System.out.println("\nMiddle node is " + myList.findMiddleNode().value);

        myList.insert(56);
        myList.printList();
        System.out.println("\nMiddle node is " + myList.findMiddleNode().value);
    }
}
